Generating target and predictor signals

ericloufoua opened the pull request:
@fissoreg 

Completed combination of clean audio and randomized noise samples, transformation of the audio to the frequency domain using the Short-Time Fourier transform (STFT), and generation of the target and predictor signals as the input for a deep learning network.

Commit:
generating target and predictor signals

Commit:
Downsampling the noise signals

fissoreg commented on preprocessing_sample_add_noise.m:
```diff
@@ -0,0 +1,39 @@
+adsNoise = audioDatastore(fullfile(noiseFolder));
+
+% Define system parameters for generating target and predictor signals
+windowLength = 256;
+win = hamming(windowLength,"periodic");
+overlap = round(0.75 * windowLength);
+ffTLength = windowLength;
+inputFs = 48e3;
+fs = 8e3;
+numFeatures = ffTLength/2 + 1;
+numSegments = 8;
```
All of this parameters are the same as in the `HelperGenerateSpeechDenoisingFeatures`. To avoid code repetition (and have a single source of truth, in this case) these should be shared across the functions.

ericloufoua replied:
I was wondering how to go about that! This was the temporary approach.

fissoreg commented on preprocessing_sample_add_noise.m:
```diff
@@ -0,0 +1,39 @@
+                              "Bandwidth",7920);
+
+% Extract Features Using Tall Arrays
+T = tall(adsTrain);
```
The Dataset is ~50GB; if you try to load it into an array, you will need ~50GB of RAM. The Tall arrays deal with this problem by keeping the data on disk and letting you access the samples "lazily", as I understand. Maybe the `gather` function below causes the memory problem?

ericloufoua replied:
I attempted to find a solution using audioDatastore and `transform `again, but when I looked over the Denoise Speech Using Deep Learning Networks example again with Samantha we realized that they used Tall Arrays specifically to process this kind of large amount of data. Here's the relevant explanation on using Tall Arrays.

![image](https://user-images.githubusercontent.com/59208173/154743336-661ec649-23ef-415a-a501-51b40be786aa.png)

I've now understood more about the result of using `gather` to evaluate the tall array, and my PC is not able to handle it even with the Parallel Computing Toolbox installed. Do you think I should also try to run tall array calculations on a cluster after installing MATLAB Parallel Server?

fissoreg replied:
I'm not sure we need parallel computations. I thought we could load a small portion of the data, work on those, save them back to disk and go on working with the next portion. This is the approach you would use to train your model, so it could work well with preprocessing too. Moreover, we probably want to do preprocessing on the fly, to avoid storing ~100s GB of preprocessed files.

But I might be wrong. Where can I find a description of the data? And of the preprocessing?

ericloufoua replied:
I think Samantha and I will look into how to perform preprocessing on the fly as you said. I'm worried about having to use such small portion of the data because we're already using a very small fraction of the original reading speech dataset (~299GB), and I'm also only trying to use a very small portion (1000/8065 files) of the currently downloaded data.

YilikaLoufoua replied:
I'm not quite sure what you're looking for in terms of a description of the data and of the preprocessing. Can you elaborate?

fissoreg reviewed (changes requested):
Hey @YilikaLoufoua  thanks for the PR! Cool stuff. I left a couple of comments on the code, but mainly we have to deal with the memory problems that you told me about. Do you have any ideas on how to make the Datastore work with the `transform` function? That might be the way to go.

fissoreg commented on preprocessing_sample_add_noise.m:
```diff
@@ -0,0 +1,39 @@
+[targets,predictors] = cellfun(@(x)HelperGenerateSpeechDenoisingFeatures(x,adsNoise,src),T,"UniformOutput",false);
+
+% Evaluate the targets and predictors.
+[targets,predictors] = gather(targets,predictors);
```
Do you know if the memory problems surface in here? That is, is the `gather` function causing the memory problems?

ericloufoua replied:
Yes! The values within the tall array is left unevaluated so that we can perform procedures on the data within, and the gather function will perform calculations to evaluate all the values. It's at this step that I start having memory issues.

Samantha-Zhan committed:
added normalization & low activity

Samantha-Zhan committed:
gitignore

Samantha-Zhan committed:
Merge branch 'preprocessing' into Data-import

Samantha-Zhan committed:
Merge pull request #13 from Samantha-Zhan/Data-import

Normalization & low activity

Commit:
Preliminary training

Training the fully connected network using 258 samples and generating relevant plots.

Commit:
Preliminary training

Training the fully connected network using 258 samples and generating relevant plots.

YilikaLoufoua commented:
Issues resolved using audioDatastore.
